synapse/rest/admin/space.py

```diff
@@ -0,0 +1,169 @@
+            user_id, [Membership.INVITE, Membership.JOIN, Membership.KNOCK]
+        )
+        user_room_ids = {room.room_id for room in user_rooms}
+
+        # Fetch the list of rooms in the space hierarchy
+        (
+            descendants,
+            inaccessible_room_ids,
+        ) = await self._space_hierarchy_handler.get_space_descendants(
+            space_id, enable_federation=include_remote_spaces
+        )
+
+        # Determine which rooms to leave by checking join rules
+        rooms_to_leave: List[str] = []
+        state_filter = StateFilter.from_types([(EventTypes.JoinRules, "")])
+        for room_id, _via in descendants:
+            if room_id not in user_room_ids:
+                # The user is not in this room. There is nothing to do here.
+                continue
+
+            current_state_ids = await self._store.get_filtered_current_state_ids(
+                room_id, state_filter
+            )
+            join_rules_event_id = current_state_ids.get((EventTypes.JoinRules, ""))
+            if join_rules_event_id is not None:
+                join_rules_event = await self._store.get_event(join_rules_event_id)
+                join_rules = join_rules_event.content.get("join_rule")
+            else:
+                # The user is invited to or has knocked on a room that is not known
+                # locally. Assume that such rooms are not public and should be left.
+                # If it turns out that the room is actually public, then we've not
+                # actually prevented the user from joining it.
+                join_rules = None
+
+            # Leave the room if it is not public, or it is the root space.
+            if join_rules != JoinRules.PUBLIC or room_id == space_id:
+                rooms_to_leave.append(room_id)
+
+        # Now start leaving rooms
+        failures: Dict[str, str] = {}
+        left_rooms: List[str] = []
+
+        fake_requester = create_requester(
+            target_user, authenticated_entity=requester.user.to_string()
+        )
+
+        for room_id in rooms_to_leave:
+            # There is a race condition here where the user may have left or been kicked
+            # from a room since their list of memberships was fetched.
+            # `update_membership` will raise if the user is no longer in the room,
+            # but it's tricky to distinguish from other failure modes.
+
+            try:
+                await self._room_member_handler.update_membership(
+                    requester=fake_requester,
+                    target=target_user,
+                    room_id=room_id,
+                    action=Membership.LEAVE,
+                    content={},
+                    ratelimit=False,
+                    require_consent=False,
+                )
+                left_rooms.append(room_id)
+            except Exception as e:
+                failures[room_id] = str(e)
```

Comment:
wondering if we want to log this exception with its traceback still so we can debug any issues that arise.

I suppose you could check the user's membership at this point to see if they were already kicked like you suggested, but it's unlikely to be worth the effort, so probably no need.